File: backend/app/api/v1/endpoints/estate_planning.py
"""
Estate Planning API Endpoints

Provides endpoints for estate planning calculations and recommendations:
- Estate tax calculations
- Trust structure recommendations
- Beneficiary optimization
- Legacy goal planning
- Gifting strategies
"""

from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime

from ....services.estate.estate_planning_service import EstatePlanningService

router = APIRouter(prefix="/api/v1/estate-planning", tags=["estate_planning"])


# Request/Response Models

class EstateTaxCalculationRequest(BaseModel):
    """Request for estate tax calculation"""
    estate_value: float = Field(..., description="Total estate value")
    state: Optional[str] = Field(None, description="Two-letter state code")
    marital_status: str = Field("single", description="Marital status")
    charitable_donations: float = Field(0.0, description="Charitable donations")
    life_insurance_value: float = Field(0.0, description="Life insurance death benefit")
    debt_value: float = Field(0.0, description="Outstanding debts")


class TrustRecommendationRequest(BaseModel):
    """Request for trust structure recommendations"""
    estate_value: float
    age: int
    marital_status: str
    has_children: bool
    charitable_intent: bool = False
    asset_protection_needs: bool = False
    business_owner: bool = False


class BeneficiaryAccount(BaseModel):
    """Account for beneficiary optimization"""
    account_id: str
    account_type: str  # ira, 401k, taxable, roth_ira, etc.
    value: float


class Beneficiary(BaseModel):
    """Beneficiary information"""
    beneficiary_id: str
    name: str
    relationship: str  # spouse, child, charity, etc.
    age: Optional[int] = None


class BeneficiaryOptimizationRequest(BaseModel):
    """Request for beneficiary optimization"""
    accounts: List[BeneficiaryAccount]
    beneficiaries: List[Beneficiary]
    estate_plan_goals: Dict[str, Any]


class LegacyGoalRequest(BaseModel):
    """Request for legacy goal calculation"""
    desired_legacy_amount: float
    current_estate_value: float
    years_to_legacy: int
    expected_return: float
    state: Optional[str] = None


class GiftingStrategyRequest(BaseModel):
    """Request for gifting strategy analysis"""
    estate_value: float
    annual_gift_amount: float
    years: int
    expected_return: float


# Dependency injection
def get_estate_service() -> EstatePlanningService:
    """Get estate planning service instance"""
    return EstatePlanningService()


# Endpoints

@router.post("/calculate-estate-tax")
async def calculate_estate_tax(
    request: EstateTaxCalculationRequest,
    service: EstatePlanningService = Depends(get_estate_service),
):
    """
    Calculate federal and state estate taxes

    Returns comprehensive estate tax analysis including:
    - Gross estate value
    - Deductions
    - Federal and state exemptions
    - Taxable amounts
    - Tax liabilities
    - Effective tax rate
    """
    try:
        result = await service.calculate_estate_tax(
            estate_value=request.estate_value,
            state=request.state,
            marital_status=request.marital_status,
            charitable_donations=request.charitable_donations,
            life_insurance_value=request.life_insurance_value,
            debt_value=request.debt_value,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/recommend-trusts")
async def recommend_trust_structures(
    request: TrustRecommendationRequest,
    service: EstatePlanningService = Depends(get_estate_service),
):
    """
    Get personalized trust structure recommendations

    Returns list of recommended trusts with:
    - Trust type and name
    - Purpose and benefits
    - Complexity level
    - Estimated setup cost
    - Suitability score
    - Priority level
    """
    try:
        recommendations = await service.recommend_trust_structures(
            estate_value=request.estate_value,
            age=request.age,
            marital_status=request.marital_status,
            has_children=request.has_children,
            charitable_intent=request.charitable_intent,
            asset_protection_needs=request.asset_protection_needs,
            business_owner=request.business_owner,
        )
        return {"recommendations": recommendations}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/optimize-beneficiaries")
async def optimize_beneficiary_designations(
    request: BeneficiaryOptimizationRequest,
    service: EstatePlanningService = Depends(get_estate_service),
):
    """
    Optimize beneficiary designations for tax efficiency

    Returns strategies for:
    - Spousal rollovers
    - Roth accounts to children
    - Charitable bequests
    - Trusts for minors
    - Contingent beneficiaries
    """
    try:
        result = await service.optimize_beneficiary_designations(
            accounts=[acc.dict() for acc in request.accounts],
            beneficiaries=[ben.dict() for ben in request.beneficiaries],
            estate_plan_goals=request.estate_plan_goals,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/calculate-legacy-goal")
async def calculate_legacy_goal(
    request: LegacyGoalRequest,
    service: EstatePlanningService = Depends(get_estate_service),
):
    """
    Calculate funding needed for legacy goal

    Returns analysis including:
    - Gross estate needed (after taxes)
    - Current vs. projected estate value
    - Shortfall or surplus
    - Required savings
    - Life insurance alternatives
    - Funding strategies
    """
    try:
        result = await service.calculate_legacy_goal(
            desired_legacy_amount=request.desired_legacy_amount,
            current_estate_value=request.current_estate_value,
            years_to_legacy=request.years_to_legacy,
            expected_return=request.expected_return,
            state=request.state,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/analyze-gifting-strategy")
async def analyze_gifting_strategy(
    request: GiftingStrategyRequest,
    service: EstatePlanningService = Depends(get_estate_service),
):
    """
    Analyze impact of annual gifting strategy

    Compares estate with and without gifting to show:
    - Future estate values
    - Estate tax with/without gifting
    - Tax savings from gifting
    - Total wealth transferred to heirs
    - Annual exclusion compliance
    """
    try:
        result = await service.analyze_gifting_strategy(
            estate_value=request.estate_value,
            annual_gift_amount=request.annual_gift_amount,
            years=request.years,
            expected_return=request.expected_return,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/health")
async def health_check():
    """Health check endpoint"""
    return {
        "status": "healthy",
        "service": "estate_planning",
        "timestamp": datetime.now().isoformat(),
    }


@router.get("/trust-types")
async def get_trust_types(
    service: EstatePlanningService = Depends(get_estate_service),
):
    """
    Get information about available trust types

    Returns details on all supported trust structures
    """
    return {"trust_types": service.TRUST_TYPES}


@router.get("/estate-tax-rates")
async def get_estate_tax_rates(
    service: EstatePlanningService = Depends(get_estate_service),
):
    """
    Get current federal and state estate tax rates

    Returns exemption amounts and tax rates by jurisdiction
    """
    return {
        "federal": {
            "exemption": service.FEDERAL_ESTATE_TAX_EXEMPTION,
            "rate": service.FEDERAL_ESTATE_TAX_RATE,
            "year": 2025,
        },
        "states": service.STATE_ESTATE_TAX,
    }

File: frontend/src/types/estatePlanning.ts
/**
 * Estate Planning Types
 *
 * TypeScript type definitions for estate planning features including:
 * - Estate tax calculations
 * - Trust structures
 * - Beneficiary designations
 * - Legacy goals
 * - Gifting strategies
 */

// Estate Tax Calculation Types

export interface EstateTaxCalculationRequest {
  estate_value: number;
  state?: string;
  marital_status: "single" | "married";
  charitable_donations?: number;
  life_insurance_value?: number;
  debt_value?: number;
}

export interface EstateTaxCalculation {
  gross_estate: number;
  deductions: number;
  taxable_estate: number;
  federal_exemption: number;
  federal_taxable_amount: number;
  federal_tax: number;
  state_exemption: number;
  state_taxable_amount: number;
  state_tax: number;
  total_tax: number;
  net_estate: number;
  effective_rate: number;
  has_federal_tax_liability: boolean;
  has_state_tax_liability: boolean;
}

// Trust Structure Types

export interface TrustRecommendationRequest {
  estate_value: number;
  age: number;
  marital_status: string;
  has_children: boolean;
  charitable_intent?: boolean;
  asset_protection_needs?: boolean;
  business_owner?: boolean;
}

export interface TrustStructure {
  name: string;
  estate_tax_benefit: boolean;
  probate_avoidance: boolean;
  asset_protection: boolean;
  complexity: "low" | "medium" | "high";
  cost: number;
  priority?: "high" | "medium" | "low";
  reason?: string;
  suitability_score?: number;
}

export interface TrustRecommendationsResponse {
  recommendations: TrustStructure[];
}

// Beneficiary Designation Types

export interface BeneficiaryAccount {
  account_id: string;
  account_type: "ira" | "401k" | "403b" | "taxable" | "roth_ira" | "roth_401k";
  value: number;
}

export interface Beneficiary {
  beneficiary_id: string;
  name: string;
  relationship: "spouse" | "child" | "charity" | "trust" | "other";
  age?: number;
}

export interface BeneficiaryOptimizationRequest {
  accounts: BeneficiaryAccount[];
  beneficiaries: Beneficiary[];
  estate_plan_goals: Record<string, any>;
}

export interface BeneficiaryStrategy {
  strategy: string;
  accounts?: string[] | string;
  beneficiary?: string;
  beneficiaries?: string[];
  reason: string;
  tax_benefit: string;
  priority: "high" | "medium" | "low";
}

export interface BeneficiaryOptimizationResponse {
  recommendations: BeneficiaryStrategy[];
  estimated_tax_savings: number;
  key_actions: string[];
}

// Legacy Goal Types

export interface LegacyGoalRequest {
  desired_legacy_amount: number;
  current_estate_value: number;
  years_to_legacy: number;
  expected_return: number;
  state?: string;
}

export interface FundingStrategy {
  strategy: string;
  description: string;
  priority: "high" | "medium" | "low";
}

export interface LegacyGoalAnalysis {
  desired_legacy: number;
  estimated_estate_tax: number;
  gross_estate_needed: number;
  current_estate_value: number;
  projected_estate_value: number;
  is_on_track: boolean;
  shortfall: number;
  surplus: number;
  annual_savings_needed: number;
  monthly_savings_needed: number;
  life_insurance_alternative: number;
  funding_strategies: FundingStrategy[];
}

// Gifting Strategy Types

export interface GiftingStrategyRequest {
  estate_value: number;
  annual_gift_amount: number;
  years: number;
  expected_return: number;
}

export interface GiftingStrategyAnalysis {
  annual_gift_amount: number;
  years_of_gifting: number;
  total_gifts: number;
  estate_without_gifting: number;
  estate_with_gifting: number;
  gifts_future_value: number;
  tax_without_gifting: number;
  tax_with_gifting: number;
  tax_savings: number;
  wealth_transferred_no_gifts: number;
  wealth_transferred_with_gifts: number;
  additional_wealth: number;
  is_within_annual_exclusion: boolean;
  gift_tax_applicable: boolean;
}

// Reference Data Types

export interface TrustTypesResponse {
  trust_types: Record<string, TrustStructure>;
}

export interface StateTaxInfo {
  exemption: number;
  rate: number;
}

export interface EstateTaxRatesResponse {
  federal: {
    exemption: number;
    rate: number;
    year: number;
  };
  states: Record<string, StateTaxInfo>;
}

// UI State Types

export interface EstatePlanningState {
  taxCalculation: EstateTaxCalculation | null;
  trustRecommendations: TrustStructure[];
  beneficiaryOptimization: BeneficiaryOptimizationResponse | null;
  legacyGoal: LegacyGoalAnalysis | null;
  giftingAnalysis: GiftingStrategyAnalysis | null;
  isLoading: boolean;
  error: string | null;
}

// Form State Types

export interface EstateTaxForm {
  estateValue: number;
  state: string;
  maritalStatus: "single" | "married";
  charitableDonations: number;
  lifeInsuranceValue: number;
  debtValue: number;
}

export interface TrustRecommendationForm {
  estateValue: number;
  age: number;
  maritalStatus: string;
  hasChildren: boolean;
  charitableIntent: boolean;
  assetProtectionNeeds: boolean;
  businessOwner: boolean;
}

export interface LegacyGoalForm {
  desiredLegacyAmount: number;
  currentEstateValue: number;
  yearsToLegacy: number;
  expectedReturn: number;
  state: string;
}

export interface GiftingStrategyForm {
  estateValue: number;
  annualGiftAmount: number;
  years: number;
  expectedReturn: number;
}

// Explicit type re-exports for Vite/TypeScript compatibility
export type {
  Beneficiary,
  BeneficiaryAccount,
  BeneficiaryOptimizationRequest,
  BeneficiaryOptimizationResponse,
  BeneficiaryStrategy,
  EstatePlanningState,
  EstateTaxCalculation,
  EstateTaxCalculationRequest,
  EstateTaxForm,
  EstateTaxRatesResponse,
  FundingStrategy,
  GiftingStrategyAnalysis,
  GiftingStrategyForm,
  GiftingStrategyRequest,
  LegacyGoalAnalysis,
  LegacyGoalForm,
  LegacyGoalRequest,
  StateTaxInfo,
  TrustRecommendationForm,
  TrustRecommendationRequest,
  TrustRecommendationsResponse,
  TrustStructure,
  TrustTypesResponse,
};
